### DISCUSSION TITLE
Linq: Skip(0) & Take(0)

### DISCUSSION THREAD

**Discussion by LeaFrock** [dotnet/runtime#47844] [Q&A] [answered]
I notice that [Queryable](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Linq.Queryable/src/System/Linq/Queryable.cs)  will still continue to create new query when Skip(0)/Take(0) is called. 
Is that necessary?  In my opinion, it should be optimized.

**Comment by huoyaoyuan** [comment]
Queryable passes the expression as-is to the provider and let the provider to decide about optimization. It shouldn't try to optimize any thing.

**Reply by LeaFrock** [reply]
> It shouldn't try to optimize any thing.

Why `shouldn't`?  Are there any cases of providers which `decide` to make `Skip(0)` a meaningful operation? Semantically speaking, it's just a no op. I prefer to think that it should be ignored by the very beginning.

Futhermore, let me take EF Core (for SQL Server) for example.

```C#
int n = ...;
var result = query.OrderBy(...).Skip(n).Take(10).ToList();
````

If `n` is `0`, it still uses `ROWINDEX` among the sql instead of `SELECT TOP(10) * from ... order by ...`.  IMO, it's better to optimize `Queryable` than let downstream providers handle it. 
However, it's just a discussion.  Please let me know if anyone has any thoughts.

**Reply by TonyValenti** [reply, marked as answer]
I can see the advantage of optimizing out those two items. Ideally, a take(0) query should never even touch a database or collection.

I kind of think that a take(0) should return enumerable.empty or whatever the queryable equivalent is.

**Reply by acaly** [reply]
As @TonyValenti mentioned, this behavior is different from the [`System.Linq.Enumerable.Linq`](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Linq/src/System/Linq/Take.cs#L18), which does not even touch the source if count is equal or less than zero (even the `GetEnumerator()` is not called)
